### PR TITLE
Review PR for Ned and Xavier

### PR DESCRIPTION
**Pull request by martynjames**
Complete review PR for xblock-google-drive

### CODE REVIEW

File: google_drive/google_calendar.py [hunk 1]
```diff
@@ -35,62 +28,92 @@ class GoogleCalendarBlock(XBlock):
...
-                width="960"
-                height="569"
-                allowfullscreen="true"
-                mozallowfullscreen="true"
-                webkitallowfullscreen="true">
-            </iframe>
-        """))
-
-    def student_view(self, context={}):
+        default="edx.org_lom804qe3ttspplj1bgeu1l3ak@group.calendar.google.com"
```
Comment (mattdrayer) [inline google_drive/google_calendar.py]: Is this a generic google account that we use for bootstrapping the component?  Should we move this value into a configuration setting?

File: google_drive/google_calendar.py [hunk 2]
```diff
@@ -35,62 +28,92 @@ class GoogleCalendarBlock(XBlock):
...
+        default=1
+    )
+
+    views = [(0, 'Week'), (1, 'Month'), (2, 'Agenda')]
+
+    # Context argument is specified for xblocks, but we are not using herein
+    def student_view(self, context):  # pylint: disable=unused-argument
         """
         Player view, displayed to the student
         """
```
Comment (mattdrayer) [inline google_drive/google_calendar.py]: Nitpick: Remove unnecessary whitespace/lines

File: google_drive/google_docs.py [hunk 3]
```diff
@@ -40,11 +35,15 @@ class GoogleDocumentBlock(XBlock):
```
Comment (mattdrayer) [inline google_drive/google_docs.py]: Nitpick: Remove unnecessary whitespace/line breaks

File: google_drive/google_docs.py [hunk 4]
```diff
@@ -40,11 +35,15 @@ class GoogleDocumentBlock(XBlock):
...
+        help=(
+            "Google provides an embed code for Drive documents. In the Google Drive document, "
+            "from the File menu, select Publish to the Web. Modify settings as needed, click "
+            "Publish, and copy the embed code into this field."
+        ),
         scope=Scope.settings,
         default=textwrap.dedent("""
             <iframe
-                src="https://docs.google.com/presentation/d/1B22AsMwG7jgE39vm1BuiJClir_JUz1q077cPEKJR2mI/embed?start=false&loop=false&delayms=3000"
+                src="https://docs.google.com/presentation/d/1x2ZuzqHsMoh1epK8VsGAlanSo7r9z55ualwQlj-ofBQ/embed?start=true&loop=true&delayms=10000"
```
Comment (mattdrayer) [inline google_drive/google_docs.py]: I assume this hash/id is a default document that we provide -- should we move this into a configuration setting so we don't have to update the code when it changes?

File: google_drive/google_calendar.py (hunk 2 above)
Comment (martynjames) [inline google_drive/google_calendar.py]: This is actually a pep8 suggested line (but not required) to have such line in place (so we possibly have tighter compliance than the rest of the code).

File: google_drive/google_docs.py [hunk 5]
```diff
@@ -54,43 +53,87 @@ class GoogleDocumentBlock(XBlock):
...
-    def studio_view(self, context):
+    # Context argument is specified for xblocks, but we are not using herein
+    def studio_view(self, context):  # pylint: disable=unused-argument
         """
         Editing view in Studio
         """
         fragment = Fragment()
-        fragment.add_content(render_template('/templates/html/google_docs_edit.html', {
+        # Need to access protected members of fields to get their default value
+        fragment.add_content(RESOURCE_LOADER.render_template('/templates/html/google_docs_edit.html', {
```
Comment (mattdrayer) [inline google_drive/google_docs.py]: I noticed that a handful of these template references are duplicated in the calendar module -- does it make sense to move them into configuration or some sort of object/property instead of having the raw strings/refs throughout the code?

File: google_drive/google_docs.py (hunk 4 above)
Comment (martynjames) [inline google_drive/google_docs.py]: Not sure how to have XBlocks use a configuration setting - not sure if I can use django settings module from within XBlock - I'll ask in chat room

File: google_drive/public/css/google_calendar.css [hunk 6]
```diff
@@ -0,0 +1,14 @@
+.google-calendar-xblock-wrapper{
```
Comment (mattdrayer) [inline google_drive/public/css/google_calendar.css]: I can't really comment much on the CSS -- perhaps @marcotuts would be a good person to review this aspect of the block?

File: google_drive/public/js/google_calendar.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/* Javascript for GoogleDocumentBlock. */
+function GoogleCalendarBlock(runtime, element) {
+
```
Comment (mattdrayer) [inline google_drive/public/js/google_calendar.js]: Unnecessary line break here?  Should we include a comment (unsure of generally-accepted JS conventions for documenting functions)

File: google_drive/public/js/google_docs_edit.js [hunk 8]
```diff
@@ -1,24 +1,117 @@
...
 
-    $('.cancel-button', element).bind('click', function() {
-        runtime.notify('cancel', {});
-    });
+    function HideAltTextInput(){
+        edit_alt_text_input.val('');
+        alt_text_item.addClass('covered');
+    }
+
+    function IsUrlValid(){
```
Comment (mattdrayer) [inline google_drive/public/js/google_docs_edit.js]: Is this function a general utility that might be useful for calendar, presentations, etc.?  Should there be a common/utils package for these types of helpers?

File: google_drive/templates/html/google_calendar_edit.html [hunk 9]
```diff
@@ -0,0 +1,68 @@
+{% load i18n %}
```
Comment (mattdrayer) [inline google_drive/templates/html/google_calendar_edit.html]: Do we need to perform an accessibility review for any HTML changes we make?  I know that @cptvitamin usually wants to see "aria" attributes added to various elements, but I'm not sure when they are actually required.

File: google_drive/tests/__init__.py [hunk 10]
```diff
@@ -0,0 +1 @@
+""" Put tests here """
```
Comment (mattdrayer) [inline google_drive/tests/__init__.py]: Maybe a more descriptive docstring here -- my first thought was 'Um, ok, so where are the tests?"

File: google_drive/tests/integration/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+""" Unit tests for google drive components """
```
Comment (mattdrayer) [inline google_drive/tests/integration/__init__.py]: Unit tests or integration tests?

File: google_drive/tests/integration/calendar_base_test.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
+""" Google Calendar integration tests """
```
Comment (mattdrayer) [inline google_drive/tests/integration/calendar_base_test.py]: Dumb question -- so by 'integration' test do we mean bok choy?  I see these are Selenium tests?

File: google_drive/tests/integration/xml/image.xml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<vertical_demo>
+    <google-document embed_code="&lt;img src='https://docs.google.com/drawings/d/1lmmxboBM5c_0WCTjhAxBdkpqQb3T8VSwtuG0TRR1ODQ/pub?w=960&amp;h=720'&gt;"/>
```
Comment (mattdrayer) [inline google_drive/tests/integration/xml/image.xml]: Is this hash something else that we might move into a settings file or constant/property/helper?

Comment (martynjames) [inline google_drive/tests/integration/xml/image.xml]: I don't think so - this is the image to use for testing. So, definitely not settings file. I like constant or property.

File: google_drive/tests/unit/test_docs.py [hunk 14]
```diff
@@ -0,0 +1,240 @@
+""" Tests for google drive components """
```
Comment (mattdrayer) [inline google_drive/tests/unit/test_docs.py]: I'm a little confused as to the different between google drive and google docs -- it seems we have a google drive component that supports google calendar and google docs and we split up most of the functionality between these two concepts.  I see calendar tests in this file -- maybe we should create a google_drive/tests/unit/test_calendar.py module to follow suit with the rest of the code?

File: google_drive/tests/integration/calendar_base_test.py (hunk 12 above)
Comment (martynjames) [inline google_drive/tests/integration/calendar_base_test.py]: Current thought seemed to be that bok choy cannot be ran on XBlock only tests, so we used selenium as suggested instead.

File: google_drive/tests/unit/test_docs.py [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        block = TestGoogleDocumentBlock.make_document_block()
+        block.usage_id = Mock()
+
+        student_fragment = block.render('student_view', Mock())
+        # pylint: disable=no-value-for-parameter
+        assert_in('<div class="google-docs-xblock-wrapper"', student_fragment.content)
+        assert_in('Google Document', student_fragment.content)
+        assert_in(
+            (
+                'https://docs.google.com/presentation/d/1x2ZuzqHsMoh1epK8VsGAlanSo7r9z55ualwQlj-ofBQ'
```
Comment (mattdrayer) [inline google_drive/tests/unit/test_docs.py]: Lots of URLs and hashes throughout this test file -- do they correspond w/ the other URLs/hashes in the code?  Definitely seems like we should push these bits of data to either config or at least a helper method, property, constant...

File: google_drive/tests/unit/test_docs.py (hunk 14 above)
Comment (martynjames) [inline google_drive/tests/unit/test_docs.py]: Me too! (a little)
Google Drive is the overall concept; Google Docs are the document types that you can store on your Google Drive. Calendars are the other things you can install on Google Drive.
(This is actually true as of early 2014, when google changed the branding / concept of what Google Drive was, it used to just be a synched-drive)

File: google_drive/tests/unit/test_docs.py [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+                '?start=true&loop=true&delayms=10000'
+            ),
+            'displayed_in': 'iframe',
+        })
+        res = block.handle('publish_event', make_request(body))
+
+        assert_equals(json.loads(res.body), {'result': 'error', 'message': 'Missing event_type in JSON data'})
+
+
+class TestGoogleCalendarBlock(unittest.TestCase):
```
Comment (mattdrayer) [inline google_drive/tests/unit/test_docs.py]: Yeah, I would definitely split this test case out into its own test_calendar.py file.  You might also consider having a TestGoogleDriveBlock parent for generalizations -- might make it easier to add support for testing new types of Drive components going forward.

File: google_drive/utils.py [hunk 17]
```diff
@@ -1,37 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (mattdrayer) [inline google_drive/utils.py]: Dropped all of the utils?  Were they not actually common/stateless units of work?

**Comment by mattdrayer** [comment]
Went through the PR -- thanks very much for setting it up -- I think I've mostly just asked a lot of possibly-dumb/irrelevant questions because I'm not too familiar with the XBlock architecture and implemenation patterns at this point -- so please bear with me :)

File: google_drive/utils.py (hunk 17 above)
Comment (martynjames) [inline google_drive/utils.py]: Yes, but they got replaced by using the stuff in xblock-utils, basically all that was left in here was the call to getLogger, which seemed silly to have it's own file

File: google_drive/templates/html/google_calendar_edit.html (hunk 9 above)
Comment (martynjames) [inline google_drive/templates/html/google_calendar_edit.html]: @cptvitamin already ran a pass through this

File: google_drive/tests/__init__.py (hunk 10 above)
Comment (martynjames) [inline google_drive/tests/__init__.py]: New pylint requires that there be a comment within this file, even though an empty version of the file is a common thing

File: google_drive/templates/html/google_calendar_edit.html (hunk 9 above)
Comment (cptvitamin) [inline google_drive/templates/html/google_calendar_edit.html]: @mattdrayer not true!  in fact, I think we add aria too much here ;)  Only add when needed.  Here is a handy guide: http://www.w3.org/TR/aria-in-html/#rec

I did take a look at this as @martynjames said, and all that is required for making an ```<iframe>``` is to include a ```title``` attribute that describes what content will be embeded.  This allows the screen reader user to quickly identify, move focus to and decide if they want to consume the embedded content or skip over it.  BTW, this is really the only time I recommend using the ```title``` attribute.

File: README.md [hunk 18]
```diff
@@ -1,2 +1,125 @@
-xblock-google-drive
-=====================
+Google Drive & Calendar XBlock [![Build Status](https://travis-ci.org/edx-solutions/xblock-google-drive.svg?branch=master)](https://travis-ci.org/edx-solutions/xblock-google-drive)
```
Comment (nedbat) [inline README.md]: A minor nit, but it would be great to be in the habit of writing README.rst files instead of README.md files.

File: README.md [hunk 19]
```diff
@@ -1,2 +1,125 @@
...
+Each time a character is added to or removed from Google Calendar ID, validation takes place.
+Analogically, validation takes place for embedded code of Google Drive File.
+
+1. Google calendar IDs are being validated against a regular expression. IDs must contain at least one '@' character, with at least one character on each side of it, ie. 'a@a'.
+
+2. Embedded code of Google Drive file is being validated on the server side, by checking the status code of the HTTP response.
+   Since error status codes start with 400, it's assumed that each status code that's larger than or equal to 400 states that file is invalid.
+   If for any reason exception occurs while getting an HTTP response, error code is returned, thus overriding default signalization that is invoked by edx platform when the 500 status code is reported.
+
+a11y
```
Comment (nedbat) [inline README.md]: Not sure people will know what this abbreviation means.

File: google_drive/google_calendar.py [hunk 20]
```diff
@@ -1,33 +1,35 @@
+"""
+Google Calendar XBlock implementation
+"""
 # -*- coding: utf-8 -*-
```
Comment (nedbat) [inline google_drive/google_calendar.py]: This line only has any effect if it's the first or second line in the file.

File: google_drive/google_calendar.py [hunk 21]
```diff
@@ -1,33 +1,35 @@
...
-# Globals ###########################################################
 
-#log = logging.getLogger(__name__)
+# Constants ###########################################################
+DEFAULT_CALENDAR_ID = "edx.org_lom804qe3ttspplj1bgeu1l3ak@group.calendar.google.com"
+CALENDAR_TEMPLATE = "/templates/html/google_calendar.html"
+CALENDAR_EDIT_TEMPLATE = "/templates/html/google_calendar_edit.html"
+CALENDAR_IFRAME = (
+    '<iframe src="https://www.google.com/calendar/embed'
+    '?mode={}&amp;src={}&amp;showCalendars=0" title="{}"></iframe>'
```
Comment (nedbat) [inline google_drive/google_calendar.py]: This template is hard to understand with positional fields. Naming them would be better.

File: google_drive/google_calendar.py [hunk 22]
```diff
@@ -35,62 +37,85 @@ class GoogleCalendarBlock(XBlock):
...
+    def student_view(self, context):  # pylint: disable=unused-argument
         """
         Player view, displayed to the student
         """
-
         fragment = Fragment()
-        context.update({
+
+        view = self.views[self.default_view][1]
+        iframe = CALENDAR_IFRAME.format(view, self.calendar_id, self.display_name)
```
Comment (nedbat) [inline google_drive/google_calendar.py]: Doesn't the display name need some HTML escaping?

Comment (nedbat) [inline google_drive/google_calendar.py]: Actually, now that I read more code, why is CALENDAR_IFRAME a separate format string? Wouldn't it be better to just include that HTML into google_calendar.html?  It's only used in this one place.

File: google_drive/google_calendar.py [hunk 23]
```diff
@@ -35,62 +37,85 @@ class GoogleCalendarBlock(XBlock):
...
+
+        fragment.add_content(RESOURCE_LOADER.render_template(CALENDAR_TEMPLATE, {
             "self": self,
```
Comment (nedbat) [inline google_drive/google_calendar.py]: This context value isn't used in the template.

File: google_drive/google_calendar.py [hunk 24]
```diff
@@ -35,62 +37,85 @@ class GoogleCalendarBlock(XBlock):
...
 
+    # suffix argument is specified for xblocks, but we are not using herein
     @XBlock.json_handler
-    def studio_submit(self, submissions, suffix=''):
-
+    def studio_submit(self, submissions, suffix=''):  # pylint: disable=unused-argument
+        """
+        Change the settings for this XBlock given by the Studio user
+        """
         self.display_name = submissions['display_name']
```
Comment (nedbat) [inline google_drive/google_calendar.py]: We should probably do something about this in the XBlock API, but you should handle failures here: submissions might not be a dict, and it might not have the keys you expect.  Not sure why, but the Schoolyourself XBlock has a bug written against it where it gets strings sometimes instead of dicts?

File: google_drive/google_docs.py [hunk 25]
```diff
@@ -1,33 +1,42 @@
...
+            <iframe
+                src="https://docs.google.com/presentation/d/1x2ZuzqHsMoh1epK8VsGAlanSo7r9z55ualwQlj-ofBQ/embed?start=true&loop=true&delayms=10000"
+                frameborder="0"
+                width="960"
+                height="569"
+                allowfullscreen="true"
+                mozallowfullscreen="true"
+                webkitallowfullscreen="true">
+            </iframe>
+        """)
```
Comment (nedbat) [inline google_drive/google_docs.py]: This "embed code" surprises me.  I know Google Docs gives an iframe tag in the embed choice, but it seems error prone to expect people to paste that in.  And the default provides a number of attributes that the Google-provided embed code does not.  Do we not want those attributes applied when the user changes the embed code?  Wouldn't it be better to take the "Link" option and turn it into an iframe ourselves?  We have no way of protecting against malformed HTML this way.

File: google_drive/google_docs.py [hunk 26]
```diff
@@ -40,57 +49,95 @@ class GoogleDocumentBlock(XBlock):
 
     embed_code = String(
         display_name="Embed Code",
-        help="Google provides an embed code for Drive documents with a variety of settings. From inside a Google Drive document, select Publish to the Web from within the File menuto get to your Embed Code with your desired settings.",
+        help=(
+            "Google provides an embed code for Drive documents. In the Google Drive document, "
+            "from the File menu, select Publish to the Web. Modify settings as needed, click "
+            "Publish, and copy the embed code into this field."
```
Comment (nedbat) [inline google_drive/google_docs.py]: You might want to mention the choice between Link and Embed.

File: google_drive/google_docs.py (hunk 25 above)
Comment (martynjames) [inline google_drive/google_docs.py]: I too was surprised, but this was considered the best MVP that we could deliver, and hope to follow up with a better experience later

File: google_drive/google_docs.py [hunk 27]
```diff
@@ -40,57 +49,95 @@ class GoogleDocumentBlock(XBlock):
...
-            </iframe>
-        """))
+        default=DEFAULT_EMBED_CODE
+    )
 
-    def student_view(self, context={}):
+    alt_text = String(
+        display_name="Alternative Text",
+        help=(
+            "In situations where image is not available to the reader, the alternative "
```
Comment (nedbat) [inline google_drive/google_docs.py]: This sentence could use some English polishing.

File: google_drive/public/js/google_calendar.js [hunk 28]
```diff
@@ -0,0 +1,17 @@
...
+        var iframe_url = $(this).attr('src');
+        $.ajax({
+            type: "POST",
+            url: runtime.handlerUrl(element, 'publish_event'),
+            data: JSON.stringify({
+                url: iframe_url,
+                displayed_in: 'iframe', event_type: 'edx.googlecomponent.calendar.displayed'
+            }),
+            success: function(){
+                $('.load_event_complete', element).val("I've published the event that indicates that the load has completed");
```
Comment (nedbat) [inline google_drive/public/js/google_calendar.js]: I haven't seen this before, what is the "I've published the event..." string used for?  I'm guessing it isn't actually displayed to the user.

Comment (nedbat) [inline google_drive/public/js/google_calendar.js]: Oh: in the integration tests.

File: pylintrc [hunk 29]
```diff
@@ -0,0 +1,328 @@
+[MASTER]
```
Comment (nedbat) [inline pylintrc]: No need for both a .pylintrc and a pylintrc file.  The travis.yml uses .pylintrc, but PyLint will find pylintrc by default.

File: requirements.txt [hunk 30]
```diff
@@ -1,3 +1,17 @@
...
+coverage==3.7
+coveralls
+django==1.4.18
+-e git+https://github.com/nosedjango/nosedjango.git@ed7d7f9aa969252ff799ec159f828eaa8c1cbc5a#egg=nosedjango-dev
+-e git+https://github.com/edx-solutions/xblock-utils.git#egg=xblock-utils
+# Normally we'd use no specific hash here, but master appears to have a defect, a proposed solution appears in the git hash below
+-e git+https://github.com/edx/xblock-sdk.git@c251b8924ef5bf89fb59a0a5df2723114cdf4154#egg=xblock-sdk
+-e git+https://github.com/edx/XBlock.git#egg=XBlock
+-e git+https://github.com/edx/bok-choy.git#egg=bok-choy
+-e git+https://github.com/pmitros/django-pyfs.git@d175715e0fe3367ec0f1ee429c242d603f6e8b10#egg=djpyfs
```
Comment (nedbat) [inline requirements.txt]: I don't see where this is used.

Comment (martynjames) [inline requirements.txt]: I believe that the workbench requires it - certainly running the tests failed on the travis servers without its presence.

File: pylintrc (hunk 29 above)
Comment (martynjames) [inline pylintrc]: Removing pylintrc file; I think this was mistakenly added anyway - we seem to have settled on .pylintrc being what we are using amongst travis-enabled projects

File: requirements.txt (hunk 30 above)
Comment (nedbat) [inline requirements.txt]: That concerns me a great deal, because xblock-sdk already mentions this repo in its setup.py install_requires settings.  I don't understand why it would be required here. This seems like travis not installing packages correctly, or perhaps xblock-sdk not being configured properly.  I guess it doesn't hurt things, because it isn't installed when installing this block.... :(

File: setup.py [hunk 31]
```diff
@@ -29,12 +29,17 @@ def package_data(pkg, roots):
     ],
     install_requires=[
         'XBlock',
+        'xblock-utils',
+    ],
+    dependency_links=[
+        'lxml',
```
Comment (nedbat) [inline setup.py]: dependency_links is meant to be a list of URLs.  This entry won't do anything.  lxml is another of those transitive needs: this XBlock doesn't use lxml.

**Comment by marjev** [comment]
@smagoun @nedbat @antoviaque 

This PR is left for review purposes. Closed PR-21 and PR-22 and implemented changes that were suggested in the comments within those PRs.

**Comment by antoviaque** [comment]
Posted a couple of answers/nits in answer to your comments on https://github.com/edx-solutions/xblock-google-drive/pull/23 but once addressed the changes made LGTM :+1:

File: google_drive/google_docs.py (hunk 27 above)
Comment (smagoun) [inline google_drive/google_docs.py]: I believe @srpearce has some updated wording

Comment (srpearce) [inline google_drive/google_docs.py]: Alternative text describes an image and appears if the image is unavailable. Screen readers for the visually impaired also use alternative text.

(Or we could just go with the first sentence - the second might be overkill.)

**Comment by marjev** [comment]
1) Changed README.md to README.rst
2) Values commonly used in tests are now contained in constants
3) Changed description text for alternative text field

**Comment by nedbat** [comment]
:+1:
